File: experiments/2026-09-19-memory-ingest/python/ingest.py
# The ingest scenarios, Python side (the duckdb wheel): one scenario per process, one CSV line out.
# Usage: python ingest.py <image-label> <file|memory> <scenario>
import sys, time, tempfile, duckdb, numpy as np
image, target, scenario = sys.argv[1:4]
def stat(key):
    for line in open("/proc/self/status"):
        if line.startswith(key): return int(line.split()[1]) / 1024
def hwm(): return stat("VmHWM")
# Writing 5 to clear_refs resets the kernel's peak-RSS counter to the current RSS, so the peak
# measured afterwards is the measured step's own, not the source frame's construction.
def reset_peak(): open("/proc/self/clear_refs", "w").write("5")
N = 50_000_000
con = duckdb.connect(tempfile.mktemp(suffix=".duckdb") if target == "file" else ":memory:")
con.execute("SET memory_limit = '300MB'")
def arrays(n=N):
    rng = np.random.default_rng(1); return rng.random(n), rng.random(n)
def create(): con.execute("CREATE TABLE t (a DOUBLE, b DOUBLE)")
def ctas(src): con.execute(f"CREATE TABLE t AS SELECT a, b FROM {src}")
def count(): return con.execute("SELECT count(*) FROM t").fetchone()[0]
base = t0 = None
def start():
    global base, t0
    import gc; gc.collect(); reset_peak(); base = stat("VmRSS"); t0 = time.time()
rows = None
if scenario in ("frame_pandas_register", "frame_pandas_register_only", "frame_pandas_append", "frame_parquet_roundtrip"):
    import pandas as pd
    a, b = arrays(); df = pd.DataFrame({"a": a, "b": b}); del a, b
    if scenario == "frame_pandas_register":
        start(); con.register("src", df); ctas("src"); rows = count()
    elif scenario == "frame_pandas_register_only":
        start(); con.register("src", df); rows = con.execute("SELECT count(*), sum(a) FROM src").fetchone()[0]
    elif scenario == "frame_pandas_append":
        create(); start(); con.append("t", df); rows = count()
    else:
        start(); f = tempfile.mktemp(suffix=".parquet"); df.to_parquet(f); ctas(f"read_parquet('{f}')"); rows = count()
elif scenario == "frame_arrow_register":
    import pyarrow as pa
    a, b = arrays(); tb = pa.table({"a": a, "b": b}); del a, b
    start(); con.register("src", tb); ctas("src"); rows = count()
elif scenario == "frame_polars_register":
    import polars as pl
    a, b = arrays(); pdf = pl.DataFrame({"a": a, "b": b}); del a, b
    start(); con.register("src", pdf); ctas("src"); rows = count()
elif scenario in ("stream_arrow_reader_generator", "stream_arrow_reader_generator_syspool"):
    import pyarrow as pa
    if scenario.endswith("_syspool"): pa.set_memory_pool(pa.system_memory_pool())
    schema = pa.schema([("a", pa.float64()), ("b", pa.float64())])
    def gen():
        rng = np.random.default_rng(1)
        for _ in range(50): yield pa.record_batch([rng.random(1_000_000), rng.random(1_000_000)], schema=schema)
    start(); con.register("src", pa.RecordBatchReader.from_batches(schema, gen())); ctas("src"); rows = count()
elif scenario == "chunks_pandas_append":
    import pandas as pd
    create(); start(); rng = np.random.default_rng(1)
    for _ in range(50): con.append("t", pd.DataFrame({"a": rng.random(1_000_000), "b": rng.random(1_000_000)}))
    rows = count()
elif scenario == "stream_stdin_csv":
    create(); start(); con.execute("COPY t FROM '/dev/stdin' (FORMAT CSV, HEADER false)"); rows = count()
elif scenario in ("stream_stdin_arrow", "stream_stdin_arrow_syspool"):
    import pyarrow as pa
    if scenario.endswith("_syspool"): pa.set_memory_pool(pa.system_memory_pool())
    start(); con.register("src", pa.ipc.open_stream(sys.stdin.buffer)); ctas("src"); rows = count()
elif scenario == "file_read_parquet":
    start(); ctas("read_parquet('/data/src.parquet')"); rows = count()
elif scenario == "file_read_csv":
    start(); ctas("read_csv('/data/src.csv')"); rows = count()
elif scenario == "engine_generate":
    start(); ctas(f"(SELECT i::DOUBLE AS a, (i * 2)::DOUBLE AS b FROM range({N}) t(i))"); rows = count()
else:
    raise SystemExit("unknown scenario " + scenario)
peak = hwm()
print(f"python,{image},{scenario},{target},{rows},{round(base)},{round(peak)},{round(peak - base)},{time.time() - t0:.1f}")

File: experiments/2026-09-19-memory-ingest/r/ingest.R
# The ingest scenarios, R side: one scenario per process, one CSV line out.
# Usage: Rscript ingest.R <image-label> <file|memory> <scenario>
args <- commandArgs(TRUE); image <- args[1]; target <- args[2]; scenario <- args[3]
stat <- function(key) { x <- readLines("/proc/self/status"); as.numeric(strsplit(x[grep(key, x)], "\\s+")[[1]][2]) / 1024 }
hwm <- function() stat("VmHWM")
# Writing 5 to clear_refs resets the kernel's peak-RSS counter to the current RSS, so the peak
# measured afterwards is the measured step's own, not the source frame's construction.
reset_peak <- function() writeLines("5", "/proc/self/clear_refs")
suppressMessages(library(DBI))
# ROWS scales the data: 50 million rows of two doubles are 800 MB; the chunked routes append a million rows at a time.
N <- as.numeric(Sys.getenv("ROWS", "50e6"))
dbdir <- if (target == "file") tempfile(fileext = ".duckdb") else ":memory:"
con <- dbConnect(duckdb::duckdb(), dbdir = dbdir); dbExecute(con, "SET memory_limit = '300MB'")
make_frame <- function(n = N) { set.seed(1); data.frame(a = runif(n), b = runif(n)) }
create <- function() dbExecute(con, "CREATE TABLE t (a DOUBLE, b DOUBLE)")
ctas <- function(from) dbExecute(con, paste0("CREATE TABLE t AS SELECT a, b FROM ", from))
base <- NA; t0 <- NA; rows <- NA
start <- function() { invisible(gc()); reset_peak(); base <<- stat("VmRSS"); t0 <<- Sys.time() }
count <- function() dbGetQuery(con, "SELECT count(*) AS n FROM t")$n
if (scenario == "frame_dbWriteTable") {
  df <- make_frame(); start(); dbWriteTable(con, "t", df); rows <- count()
} else if (scenario == "frame_dbAppendTable") {
  df <- make_frame(); create(); start(); dbAppendTable(con, "t", df); rows <- count()
} else if (scenario == "frame_register_only") {
  df <- make_frame(); start(); duckdb::duckdb_register(con, "src", df)
  rows <- dbGetQuery(con, "SELECT count(*) AS n, sum(a) AS s FROM src")$n
} else if (scenario == "frame_arrow_register") {
  df <- make_frame(); start(); tb <- arrow::as_arrow_table(df); duckdb::duckdb_register_arrow(con, "src", tb); ctas("src"); rows <- count()
} else if (scenario == "frame_nanoarrow_register") {
  df <- make_frame(); start()
  rdr <- arrow::as_record_batch_reader(nanoarrow::as_nanoarrow_array_stream(df))
  duckdb::duckdb_register_arrow(con, "src", rdr); ctas("src"); rows <- count()
} else if (scenario == "frame_nanoparquet_roundtrip") {
  df <- make_frame(); start(); f <- tempfile(fileext = ".parquet"); nanoparquet::write_parquet(df, f)
  ctas(sprintf("read_parquet('%s')", f)); rows <- count()
} else if (scenario == "chunks_dbAppendTable") {
  create(); start(); for (i in seq_len(N / 1e6)) dbAppendTable(con, "t", make_frame(1e6)); rows <- count()
} else if (scenario == "stream_stdin_csv") {
  create(); start(); dbExecute(con, "COPY t FROM '/dev/stdin' (FORMAT CSV, HEADER false)"); rows <- count()
} else if (scenario %in% c("stream_stdin_arrow", "stream_stdin_arrow_1thread")) {
  if (scenario == "stream_stdin_arrow_1thread") dbExecute(con, "SET threads = 1")
  start(); rdr <- arrow::as_record_batch_reader(nanoarrow::read_nanoarrow(file("stdin", "rb")))
  duckdb::duckdb_register_arrow(con, "src", rdr); ctas("src"); rows <- count()
} else if (scenario == "stream_stdin_arrow_dbAppendTableArrow") {
  # DBI's default: pulls one batch at a time on R's thread and appends each as a data frame.
  create(); start(); DBI::dbAppendTableArrow(con, "t", nanoarrow::read_nanoarrow(file("stdin", "rb"))); rows <- count()
} else if (startsWith(scenario, "dataset_arrow_to_duckdb")) {
  # What arrow::to_duckdb() does: the arrow package's dataset scanner reads the file, the engine ingests the stream.
  if (endsWith(scenario, "_1thread")) invisible(dbExecute(con, "SET threads = 1"))
  start(); duckdb::duckdb_register_arrow(con, "src", arrow::open_dataset("/data/src.parquet"))
  if (endsWith(scenario, "_count")) rows <- dbGetQuery(con, "SELECT count(*) AS n FROM src")$n else { ctas("src"); rows <- count() }
} else if (scenario == "file_read_parquet") {
  start(); ctas("read_parquet('/data/src.parquet')"); rows <- count()
} else if (scenario == "file_read_csv") {
  start(); ctas("read_csv('/data/src.csv')"); rows <- count()
} else if (scenario == "file_duckdb_read_csv") {
  start(); duckdb::duckdb_read_csv(con, "t", "/data/src.csv"); rows <- count()
} else if (scenario == "engine_generate") {
  start(); ctas(sprintf("(SELECT i::DOUBLE AS a, (i * 2)::DOUBLE AS b FROM range(%d) t(i))", as.integer(N))); rows <- count()
} else stop("unknown scenario ", scenario)
peak <- hwm()
cat(sprintf("r,%s,%s,%s,%d,%d,%d,%d,%.1f\n", image, scenario, target, as.integer(rows), round(base), round(peak), round(peak - base), as.numeric(Sys.time() - t0, units = "secs")))
dbDisconnect(con, shutdown = TRUE)
